urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/driver/examples/INDIDriverExample.java

```diff
@@ -0,0 +1,156 @@
+  }
+
+  // No BLOB Properties can be set by clients. Empty method  
+  @Override
+  public void processNewBLOBValue(INDIBLOBProperty property, Date timestamp, INDIBLOBElementAndValue[] elementsAndValues) {
+  }
+
+  // No Text Properties can be set by clients. Empty method
+  @Override
+  public void processNewTextValue(INDITextProperty property, Date timestamp, INDITextElementAndValue[] elementsAndValues) {
+  }
+
+  // Processes the changes sent by the client to the Switch Properties. If the Switch property is the CONNECTION one it adds or removes the image and send Properties. If the Property is the "sendImage", it loads an image from disk, sets it to the image property and sends it back to the client.
+  @Override
+  public void processNewSwitchValue(INDISwitchProperty property, Date timestamp, INDISwitchElementAndValue[] elementsAndValues) {
+
+    if (property == sendP) {  // If the property is the sendImage one
+      if (elementsAndValues.length > 0) {  // If any element has been updated
+        INDISwitchElement el = elementsAndValues[0].getElement();
+        SwitchStatus s = elementsAndValues[0].getValue();
+
+        if ((el == sendE) && (s == SwitchStatus.ON)) {  // If the sendImage element has been switched one we send the image
+          boolean imageLoaded = loadImageFromFile();
+
+          if (imageLoaded) {
+            sendP.setState(PropertyStates.OK);  // Set the state of the sendImage property as OK
+
+            imageP.setState(PropertyStates.OK); // Set the state of the image property as OK
+
+            try {
+              updateProperty(sendP); // Send the sendImage property to the client.
+              updateProperty(imageP); // Send the image property to the client.
+            } catch (INDIException e) {
+              e.printStackTrace();
+              System.exit(-1);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Add the image and send properties changes
+  @Override
+  public void driverConnect(Date timestamp) {
+    printMessage("Driver connect");
+    this.addProperty(imageP);
+    this.addProperty(sendP);
+  }
+
+  // Remove the image and send properties changes
+  @Override
+  public void driverDisconnect(Date timestamp) {
+    printMessage("Driver disconnect");
+    this.removeProperty(imageP);
+    this.removeProperty(sendP);
+  }
+
+  // Loads the image "image.jpg" from the same directory into the image property. Returns true if the loading has been succesful. false otherwise.
+  private boolean loadImageFromFile() {
+    if (imageE.getValue().getSize() == 0) { // If it has not been already loaded
+
+      byte[] fileContents;
+
+      try {
+        File file = new File("image.jpg");
+
+// Create a buffer big enough to hold the file
+        int size = (int) file.length();
+        fileContents = new byte[size];
+// Create an input stream from the file object and read it all
+        FileInputStream in = new FileInputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at line 140. The resource is closed at line 141. There are other execution paths that do not contain closure statements, for example, when FileInputStream.read throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).